### PR TITLE
libct/cgroups: support Cgroups.Resources.Unified

### PR DESCRIPTION
**Pull request by kolyshkin**
Add support for unified resource map (as per [1]), and add some test
cases for the new functionality.
    
[1] https://github.com/opencontainers/runtime-spec/pull/1040

Fixes: #2563

The biggest issue, I think, is that the mapping of the "unified" parameters
to systemd properties is not implemented, meaning there can be cases
where systemd unit settings and in-kernel cgroup settings won't agree.

I don't think it can cause any issues, but it's just doesn't look right.

### CODE REVIEW

**Comment by kolyshkin** [comment]
That intermittent download failure, again
```
 ---> b2ec3d2c2048
Step 13/23 : RUN curl -o /usr/local/bin/umoci -fsSL https://github.com/opencontainers/umoci/releases/download/v0.4.5/umoci.amd64     && chmod +x /usr/local/bin/umoci
 ---> Running in b2823a13a789
curl: (22) The requested URL returned error: 429 too many requests
The command '/bin/sh -c curl -o /usr/local/bin/umoci -fsSL https://github.com/opencontainers/umoci/releases/download/v0.4.5/umoci.amd64     && chmod +x /usr/local/bin/umoci' returned a non-zero code: 22
Makefile:62: recipe for target 'runcimage' failed
```

Wonder if there's a better alternative. For some reason, this never happens to any other URLs downloaded from github

**Comment by kolyshkin** [comment]
Rebased

File: libcontainer/integration/exec_test.go [hunk 1]
```diff
@@ -705,6 +705,150 @@ func testRunWithKernelMemory(t *testing.T, systemd bool) {
...
+}
+
+func TestCgroupResourcesUnifiedSystemd(t *testing.T) {
+	if !systemd.IsRunningSystemd() {
+		t.Skip("Systemd is unsupported")
+	}
+	testCgroupResourcesUnified(t, true)
+}
+
+func testCgroupResourcesUnified(t *testing.T, systemd bool) {
```
Comment (AkihiroSuda) [inline libcontainer/integration/exec_test.go:754]: We should also have integration tests

Comment (kolyshkin) [reply libcontainer/integration/exec_test.go:754]: Sure, will do

Comment (kolyshkin) [reply libcontainer/integration/exec_test.go:754]: Added a few basic tests. Let me know if you had something else in mind.

**Comment by kolyshkin** [comment]
@giuseppe @mrunalp PTAL

File: tests/integration/cgroups.bats [hunk 2]
```diff
@@ -176,3 +178,37 @@ function setup() {
...
+}
+
+@test "runc run (cgroup v2 + unified resources)" {
+    requires root cgroups_v2
+
+    set_cgroups_path "$BUSYBOX_BUNDLE"
+    update_config ' .linux.resources.memory |= {"limit": 33554432}
+                  | .linux.resources.memorySwap |= {"limit": 33554432}
+                  | .linux.resources.unified |=
+                      {"memory.min": "131072", "memory.max": "10485760" }' \
```
Comment (AkihiroSuda) [inline tests/integration/cgroups.bats:201]: We should also test:
* `pids.max` (which doesn't conflict v1-compat config)
* `non.existent.value`
* `../../../etc/shadow`

File: tests/integration/cgroups.bats [hunk 3]
```diff
@@ -176,3 +178,37 @@ function setup() {
...
+    set_cgroups_path "$BUSYBOX_BUNDLE"
+    set_resources_limit "$BUSYBOX_BUNDLE"
+    update_config '.linux.resources.unified |= {"memory.min": "131072"}' "$BUSYBOX_BUNDLE"
+
+    runc run -d --console-socket "$CONSOLE_SOCKET" test_cgroups_unified
+    [ "$status" -ne 0 ]
+    [[ "$output" == *'invalid configuration'* ]]
+}
+
+@test "runc run (cgroup v2 + unified resources)" {
```
Comment (AkihiroSuda) [inline tests/integration/cgroups.bats:194]: Can we also have test for pure-unified config

File: tests/integration/cgroups.bats (hunk 2 above)
Comment (kolyshkin) [reply tests/integration/cgroups.bats:201]: I am already testing the last two in the libcontainer/integration tests. Not sure how adding these cases to bats will add any value.

Not sure what you mean by "v1-compat config". The runtime-spec says 'resources.unified' can't be used on v1, and there's a code to enforce it (checking if "unified" is set, doesn't look into individual parameters). Do you mean you want to test this works if we set "pids.max"?

Comment (AkihiroSuda) [reply tests/integration/cgroups.bats:201]: All the current fields in `.linux.resources.unified` overrides the v1 config (`.linux.resources.{memory,memorySwap}`).

We should ensure that we can set unified fields that do not override v1.

Comment (kolyshkin) [reply tests/integration/cgroups.bats:201]: There is a code that rejects config if `.linux.resources.unified` set for cgroupv1, here:
https://github.com/opencontainers/runc/pull/2584/commits/b14556f2678eb12cbacdc12ae3ddf6dc952e3fbe#diff-252ac01f71681e9ed8e7ce0571cdb6acR349-R351
and here:
https://github.com/opencontainers/runc/pull/2584/commits/b14556f2678eb12cbacdc12ae3ddf6dc952e3fbe#diff-bf0fd9d73921163a3f447b24a8187ee7

I check that this code works in two tests, here:
https://github.com/opencontainers/runc/pull/2584/commits/b14556f2678eb12cbacdc12ae3ddf6dc952e3fbe#diff-bd87cca3665e082adb4ba55ee22e82b5
and here:
https://github.com/opencontainers/runc/pull/2584/commits/d0397b925803e3cc78e79ec7ef1da622529595c1#diff-1be821a9fd1920ed1cd1090d8a6bea39R182

The rejection code does not look into what's set under the `unified`, so it does not make sense to add some specific parameters in there -- anything is rejected.

File: tests/integration/cgroups.bats (hunk 3 above)
Comment (kolyshkin) [reply tests/integration/cgroups.bats:194]: will do

Comment (kolyshkin) [reply tests/integration/cgroups.bats]: added; PTAL

**Comment by kolyshkin** [comment]
Added bats test with more unified parameters (using only `resources.unified`), and libct/int test case using read-only file (`pids.current`) as a key to check for error.

**Review by AkihiroSuda — DISMISSED** [review]

**Comment by AkihiroSuda** [comment]
@giuseppe PTAL?

File: libcontainer/cgroups/fs2/fs2.go [hunk 4]
```diff
@@ -206,10 +209,42 @@ func (m *manager) Set(container *configs.Config) error {
...
 }
 
+func (m *manager) setUnified(res map[string]string) error {
+	for k, v := range res {
+		if strings.Contains(k, "/") {
+			return fmt.Errorf("unified resource %q must be a file name (no slashes)", k)
+		}
+		if err := fscommon.WriteFile(m.dirPath, k, v); err != nil {
+			errC := errors.Cause(err)
+			// cgroup2 returns EPERM (rather than ENOENT) for
```
Comment (giuseppe) [inline libcontainer/cgroups/fs2/fs2.go:226]: is EPERM because the file is opened with `O_CREAT`?

**Review by giuseppe — COMMENTED** [review]
just a question, otherwise LGTM

File: libcontainer/cgroups/fs2/fs2.go (hunk 4 above)
Comment (kolyshkin) [reply libcontainer/cgroups/fs2/fs2.go:226]: Right. `fscommon.WriteFile` use `ioutil.WriteFile` which calls `os.OpenFile(filename, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, perm)`.

Let me fix the comment.

Wonder if we should not use `O_CREAT` as it makes no sense for cgroupfs.

Comment (kolyshkin) [reply libcontainer/cgroups/fs2/fs2.go]: fixed

**Comment by giuseppe** [comment]
LGTM

File: libcontainer/cgroups/fs2/fs2.go [hunk 5]
```diff
@@ -206,10 +209,41 @@ func (m *manager) Set(container *configs.Config) error {
...
+					return fmt.Errorf("unified resource %q must be in the form CONTROLLER.PARAMETER", k)
+				}
+				c := sk[0]
+				if _, ok := m.controllers[c]; !ok && c != "cgroup" {
+					return fmt.Errorf("unified resource %q can't be set: controller %q not available", k, c)
+				}
+			}
+			return errors.Wrapf(err, "can't set unified resource %q", k)
+		}
+
```
Comment (mrunalp) [inline libcontainer/cgroups/fs2/fs2.go:241]: nit: space

**Review by mrunalp — APPROVED** [review]
We need follow up for systemd but this looks like a good start. Thanks!

File: libcontainer/cgroups/fs2/fs2.go (hunk 5 above)
Comment (kolyshkin) [reply libcontainer/cgroups/fs2/fs2.go]: fixed

**Review by mrunalp — APPROVED** [review]

**Comment by mrunalp** [comment]
@AkihiroSuda ptal

**Review by AkihiroSuda — APPROVED** [review]

**Comment by AkihiroSuda** [comment]
@kolyshkin Could you open an issue for tracking systemd issue?

**Comment by kolyshkin** [comment]
> Could you open an issue for tracking systemd issue?

* https://github.com/opencontainers/runc/issues/2602
* https://github.com/opencontainers/runc/issues/2603

I'm not quite sure which way to go -- let's discuss in there

**Pull request merged**